kv: only scan intents span for QueryIntent request

andrewbaptist opened the pull request:
Change QueryIntent to only read from the lock table. Previously the request required merging the read from the MVC iterator with the lock table. This should improve performance.

Also, add unit testing for QueryIntent

Fixes #69716

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81830)
<!-- Reviewable:end -->

Commit:
kvserver: fix incorrect comment information

Release note: None

Commit:
kv: only scan intents span for QueryIntent request

Change QueryIntent to only read from the lock table.
Previously the request required merging the read from the MVCC
iterator with the lock table. This should improve performance.

Also, add unit testing for QueryIntent.

Fixes #69716

Release note: None

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N2wYhi69u5zNbrXC__5:-N2wYhi70ifamTRy29ff:bfyj7wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7cdc7512982d0aedce3d4159c2f2c974de334/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L116)):*
> ```Go
> 			return result.Result{}, err
> 		}
> 		// This should not be possible. There can only be one outstanding write
> ```

We're going to relax this sometime in the future, and I would like to hide such assumptions inside the storage package. So consider moving all the real work here into a `func QueryIntent(ctx context.Context, reader Reader, ...`, say in engine.go.


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N2wYhi69u5zNbrXC__5:-N3-OACo2cLrsAmSuClg:b-bgunsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7cdc7512982d0aedce3d4159c2f2c974de334/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L116)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We're going to relax this sometime in the future, and I would like to hide such assumptions inside the storage package. So consider moving all the real work here into a `func QueryIntent(ctx context.Context, reader Reader, ...`, say in engine.go.
</blockquote></details>

Moved this to engine.go


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
*[`pkg/storage/engine.go` line 891 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-Ubg2EEEOH4VpwIqw:-N3-Ubg33EPmY3Xvh4Ht:bcsz3qd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L891)):*
> ```Go
> // returned from the iterator unable to unmarshal the intent more than one entry
> // for this key
> func QueryIntent(reader Reader, key roachpb.Key) (*roachpb.Intent, error) {
> ```

I don't love the return value signature of this method. Is there a better approach to handle this. There are three possible returns:
error - something failed
nil intent - no intent found
valid intent - intent found

In something like Rust it could be written as Result<Option<Intent>, Err>, but not sure the best practice for go. 


<!-- Sent from Reviewable.io -->

knz commented:
In go that would be `(hasIntent bool, intent *Intent, err error)`

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/storage/engine.go` line 921 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-VFx69zTuD2WytZbt:-N3-VFx69zTuD2WytZbu:b-eyraxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L921)):*
> ```Go
> 			return intent, err
> 		}
> 		// TODO: Why can't this be inlined?
> ```

How should this be done? I tried a few variations of inlining this, but nothing seemed to work correctly


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 404 files at r1, 3 of 3 files at r4, 3 of 4 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 54 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N304QMrCmKdSe3XOFYN:-N304QMrCmKdSe3XOFYO:b-bdx36f) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L54)):*
> ```Go
> 	ctx context.Context, reader storage.Reader, cArgs CommandArgs, resp roachpb.Response,
> ) (result.Result, error) {
> ```

nit: stray line.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 78 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N307gjmE_I46kLpuXUn:-N307gjmE_I46kLpuXUo:b-2xf4x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L78)):*
> ```Go
> 
> 	// Read from the lock table to see if an intent exists.
> 	// Iterate over the lock key space with this key as a lower bound.
> ```

This comment is stale. Consider moving to `storage/engine.go`
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 101 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N307owWEyWrEGEppoel:-N307owWEyWrEGEppoem:b-g58wen) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L101)):*
> ```Go
> 		// past its expected timestamp.
> 		if !reply.FoundIntent {
> 			log.Infof(ctx, "intent mismatch requires - %v == %v and %v == %v and %v <= %v",
> ```

This feels like it could be noisy to me. There are cases where an intent not being found is valid. Consider putting this behind a log verbosity level, like `log.VEventf(ctx, 2, ...)`. That way, it will show up in traces and also logs if logging verbosity is dialed up sufficiently for this file.
___
*[`pkg/storage/engine.go` line 891 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-Ubg2EEEOH4VpwIqw:-N305LYz3MOHU-ufE-Pe:b-kygv1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L891)):*
> In go that would be (hasIntent bool, intent *Intent, err error)

I don't think that's quite right. We wouldn't return a pointer to an Intent if we weren't using nullability to indicate whether the intent existed or not. We have two options here, either of which is idiomatic in Go:
- `(intent *Intent, err error)`
- `(ok bool, intent Intent, err error)`
___
*[`pkg/storage/engine.go` line 921 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-VFx69zTuD2WytZbt:-N3069hV5sH0a2yCeZbZ:b-bu8w6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L921)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

How should this be done? I tried a few variations of inlining this, but nothing seemed to work correctly
</blockquote></details>

If you want a pointer to an intent then you are going to have to do something like this. Go won't let you take the address of a return value directly. You first need to assign it to a variable on the stack, then take the address of that, which will force it to escape.

Do we need this? `var intent *roachpb.Intent = nil` isn't giving you anything. Couldn't we do something like:
```
intent := roachpb.MakeIntent(meta.Txn, checkKey)
return &intent, nil
```
___
*[`pkg/storage/engine.go` line 886 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N304eyq781oxAEFp-LY:-N304eyq781oxAEFp-LZ:bwsc8tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L886)):*
> ```Go
> }
> 
> // QueryIntent will look up an intent given a key. It there is no intent for a
> ```

nit: to parallel `ScanIntents`, I think this should be called `GetIntent`, not `QueryIntent`.
___
*[`pkg/storage/engine.go` line 903 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N306TwIAdNkKIuzQWqI:-N306TwJ4SXgqOc4lYQi:bc8ej5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L903)):*
> ```Go
> 		return intent, err
> 	}
> 	if valid {
> ```

style nit: Go prefers early returns and minimal outdenting. So instead of:
```go
if valid {
    ... long ...
}
return intent, nil
```
prefer:
```go
if !valid {
    return nil, nil
}
... long ...
```
___
*[`pkg/storage/engine.go` line 920 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3078gjE8qARRhOOSIu:-N3078gjE8qARRhOOSIv:bqqsyj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L920)):*
> ```Go
> 		if err = protoutil.Unmarshal(iter.UnsafeValue(), &meta); err != nil {
> 			return intent, err
> 		}
> ```

We'll want to validate that `meta.Txn != nil`, or else throw an `errors.AssertionFailedf`.
___
*[`pkg/storage/engine.go` line 928 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N307LI70um2kAyFBBOq:-N307LI70um2kAyFBBOr:b8lwqiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L928)):*
> ```Go
> 		if err != nil {
> 			// we expect false, but not an error
> 			return intent, err
> ```

Generally in Go you shouldn't return a valid value and an error. So here and below, prefer `return nil, err`.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 45 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N308L756E-I5RdhUTe3:-N308L756E-I5RdhUTe4:b-9gp5j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L45)):*
> ```Go
> 	}
> 
> 	// write three keys in three separate transactions
> ```

(throughout this test) Go's style guide prefers proper capitalization and punctuation in comments: https://go.dev/doc/effective_go#commentary.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N308jxzCxwC7NRcD7iN:-N308jxzCxwC7NRcD7iO:b2prnpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L114)):*
> ```Go
> 			}
> 			cArgs.EvalCtx = evalCtx.EvalContext()
> 			var resp roachpb.QueryIntentResponse
> ```

The test never seems to be consulting `QueryIntentResponse.FoundIntent`. So if `!test.errorFlag`, it's not really testing anything. Perhaps this is where your question came from on Slack.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, and @sumeerbhola)

___
*[`pkg/storage/engine.go` line 886 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N30ALsj-oncXIHnZa5E:-N30ALsj-oncXIHnZa5F:b-lo32kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L886)):*
> ```Go
> }
> 
> // QueryIntent will look up an intent given a key. It there is no intent for a
> ```

Should we add unit testing around this function as well? Similar to `TestScanIntents`.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
OK, thanks - in this case, the purpose of the *Intent was to cover the hasIntent or not, so I think it is OK.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-:-N30NyZt6HwmJVEgJYRU:b-inzny7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N2wYhi69u5zNbrXC__5:-N30JsVn1qeFRqb0QLqF:bm8x24e) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c7cdc7512982d0aedce3d4159c2f2c974de334/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L116)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Moved this to engine.go
</blockquote></details>

OK, moved
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 54 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N304QMrCmKdSe3XOFYN:-N30EwqE7y1CzvoQxVnf:blndajj) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L54)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: stray line.
</blockquote></details>

OK, Fixed
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 101 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N307owWEyWrEGEppoel:-N30EtNu0ARykwh8ISy7:biy8jdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L101)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This feels like it could be noisy to me. There are cases where an intent not being found is valid. Consider putting this behind a log verbosity level, like `log.VEventf(ctx, 2, ...)`. That way, it will show up in traces and also logs if logging verbosity is dialed up sufficiently for this file.
</blockquote></details>

OK, Changed
___
*[`pkg/storage/engine.go` line 891 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-Ubg2EEEOH4VpwIqw:-N30FLBS8qfaAxoDA9FL:bvwu2cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L891)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> In go that would be (hasIntent bool, intent *Intent, err error)

I don't think that's quite right. We wouldn't return a pointer to an Intent if we weren't using nullability to indicate whether the intent existed or not. We have two options here, either of which is idiomatic in Go:
- `(intent *Intent, err error)`
- `(ok bool, intent Intent, err error)`
</blockquote></details>

OK, leaving as is - this is the reason I returned the pointer, so just wanted to make sure it was idiomatic
___
*[`pkg/storage/engine.go` line 921 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3-VFx69zTuD2WytZbt:-N30GacO3L7GQXtTOShR:brujzsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a81703fad534097ce2d2c4db6dd83aa265c0a4f/pkg/storage/engine.go#L921)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If you want a pointer to an intent then you are going to have to do something like this. Go won't let you take the address of a return value directly. You first need to assign it to a variable on the stack, then take the address of that, which will force it to escape.

Do we need this? `var intent *roachpb.Intent = nil` isn't giving you anything. Couldn't we do something like:
```
intent := roachpb.MakeIntent(meta.Txn, checkKey)
return &intent, nil
```
</blockquote></details>

OK, thanks!
___
*[`pkg/storage/engine.go` line 886 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N304eyq781oxAEFp-LY:-N30FApu4X962kbqa9JO:b-8kurb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L886)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: to parallel `ScanIntents`, I think this should be called `GetIntent`, not `QueryIntent`.
</blockquote></details>

OK, changed
___
*[`pkg/storage/engine.go` line 886 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N30ALsj-oncXIHnZa5E:-N30GoLf0cSoLpErjUIg:b-s6xgqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L886)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we add unit testing around this function as well? Similar to `TestScanIntents`.
</blockquote></details>

Working, I should have moved the testing after splitting this function into engine.go
___
*[`pkg/storage/engine.go` line 903 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N306TwIAdNkKIuzQWqI:-N30G87X5Psq1oIDXvFv:b-eyu65x) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L903)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

style nit: Go prefers early returns and minimal outdenting. So instead of:
```go
if valid {
    ... long ...
}
return intent, nil
```
prefer:
```go
if !valid {
    return nil, nil
}
... long ...
```
</blockquote></details>

OK, changed - also changed to use nil, err in most places to make it more clear
___
*[`pkg/storage/engine.go` line 920 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3078gjE8qARRhOOSIu:-N30HQNzC2AgYYwoCtGO:b-9a27f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L920)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We'll want to validate that `meta.Txn != nil`, or else throw an `errors.AssertionFailedf`.
</blockquote></details>

OK, added check
___
*[`pkg/storage/engine.go` line 928 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N307LI70um2kAyFBBOq:-N30HWOy8e_1rM36_c0F:brhzkdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L928)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Generally in Go you shouldn't return a valid value and an error. So here and below, prefer `return nil, err`.
</blockquote></details>

OK, changed all to make this more clear
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 45 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N308L756E-I5RdhUTe3:-N30LIc40zlMMkV8wWNv:b-oodjga) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L45)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

(throughout this test) Go's style guide prefers proper capitalization and punctuation in comments: https://go.dev/doc/effective_go#commentary.
</blockquote></details>

OK, updated
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N308jxzCxwC7NRcD7iN:-N30J1POConL8lAE7n_7:bixhuh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L114)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The test never seems to be consulting `QueryIntentResponse.FoundIntent`. So if `!test.errorFlag`, it's not really testing anything. Perhaps this is where your question came from on Slack.
</blockquote></details>

Working, I missed checking the response object that comes back.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/storage/engine.go` line 886 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N30ALsj-oncXIHnZa5E:-N312dFbBL1TM-5g4hMe:b-2qayrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/storage/engine.go#L886)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Working, I should have moved the testing after splitting this function into engine.go
</blockquote></details>

OK, added unit test that at least cover the main cases (found/not found). Lets chat if there are easy ways to test a lot of the other error cases. 
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 114 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N308jxzCxwC7NRcD7iN:-N312sMz98SlyMwbNv2K:bj1eeoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/80f1337f629cb6324f22bd664c2332d2512f54f2/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L114)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Working, I missed checking the response object that comes back.
</blockquote></details>

OK - I added additional checks against the response object. This covers the main case of the reply.FoundIntent, but doesn't correctly check the WriteTimestamp on the reply. I could probably add that if you think it is worthwhile to do. It gets set in the one case of the transaction write timestamp is less than the intent write timestamp.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
This is getting close. Nice job!

> Reviewed 3 of 3 files at r7, 5 of 5 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-:-N35MKVy1HXAHA7AL9m3:bxxsgb6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 87 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35L0b56sSHgP6M9GsY:-N35L0b56sSHgP6M9GsZ:b-p1310u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L87)):*
> ```Go
> 
> 	if intentPtr != nil {
> 		intent := *intentPtr
> ```

This doesn't feel necessary. Consider leaving `intentPtr` called `intent` and never explicitly dereferencing. Go implicitly dereferences when accessing fields and methods.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 121 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35L8CM1JbMY5WN6Idb:-N35L8CM1JbMY5WN6Idc:b-4awql2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L121)):*
> ```Go
> 				if ownTxn {
> 					reply.Txn = h.Txn.Clone()
> 					reply.Txn.WriteTimestamp.Forward((intent).Txn.WriteTimestamp)
> ```

nit: why the parenthesis?
___
*[`pkg/storage/engine.go` line 891 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35LFQP7hXSdvY7MVhV:-N35LFQP7hXSdvY7MVhW:b2garp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine.go#L891)):*
> ```Go
> // intent, missing transaction on the intent or multiple intents for this key.
> func GetIntent(reader Reader, key roachpb.Key) (*roachpb.Intent, error) {
> 	// translate to a key in the lock space
> ```

Same point about the the capitalization and punctuation in comments.
___
*[`pkg/storage/engine.go` line 923 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35KrPlBoL1ZdXUFavA:-N35KrPlBoL1ZdXUFavB:b-trsmob) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine.go#L923)):*
> ```Go
> 	}
> 	if meta.Txn == nil {
> 		return nil, errors.AssertionFailedf("Txn is null for key %v, intent %v", key, meta)
> ```

nit: error messages are generally lower-case
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 128 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35MH7z5OwwVoiAmwVt:-N35MH7z5OwwVoiAmwVu:b-mez5pz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L128)):*
> ```Go
> 			switch test.response {
> 			case Error:
> 				require.Error(t, err)
> ```

Should we assert on the contents of the error?
___
*[`pkg/storage/engine_test.go` line 1636 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35M0-y3SNgpAXB6W09:-N35M0-y3SNgpAXB6W0A:bqz2631) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine_test.go#L1636)):*
> ```Go
> 				require.NoError(t, err)
> 				if test.found {
> 					require.NotNil(t, intent)
> ```

Let's also test the contents of the intent. The txn should be correct (maybe just check the `txn.ID`) and the key should be correct.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 87 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35L0b56sSHgP6M9GsY:-N35UXSPARQ9ZrDEqFqG:b-bevuc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L87)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This doesn't feel necessary. Consider leaving `intentPtr` called `intent` and never explicitly dereferencing. Go implicitly dereferences when accessing fields and methods.
</blockquote></details>

OK, fixed - I didn't realize about the implict deference - but makes this much cleaner
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 121 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35L8CM1JbMY5WN6Idb:-N35UizJ1NF3sttkNs3Z:b-j4jgnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: why the parenthesis?
</blockquote></details>

OK removed
___
*[`pkg/storage/engine.go` line 891 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35LFQP7hXSdvY7MVhV:-N35V3caAdZGYeX2x-wo:bm4ypj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine.go#L891)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same point about the the capitalization and punctuation in comments.
</blockquote></details>

OK, fixed
___
*[`pkg/storage/engine.go` line 923 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35KrPlBoL1ZdXUFavA:-N35Vf6j06e1gF4jb7w9:b-8kurb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine.go#L923)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: error messages are generally lower-case
</blockquote></details>

OK, changed
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 128 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35MH7z5OwwVoiAmwVt:-N3689kv8MhGZHR3TKDA:b-tc7i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L128)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we assert on the contents of the error?
</blockquote></details>

OK, I added a check on the type of the errors - it is a little hacky though, so if you see a better way to do this I can change it.
___
*[`pkg/storage/engine_test.go` line 1636 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35M0-y3SNgpAXB6W09:-N3680UGDt2eJGFGHSZk:b3dhtpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/storage/engine_test.go#L1636)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's also test the contents of the intent. The txn should be correct (maybe just check the `txn.ID`) and the key should be correct.
</blockquote></details>

OK, Good idea - added (and check key to be safe)


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> other than the proposed cleanup to avoid reflection in `cmd_query_intent_test.go`.

> Reviewed 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-:-N3QBInX8Hf6dAk41uDC:bpfqeho)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 128 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35MH7z5OwwVoiAmwVt:-N3QAocpCUoB-rkRBIBl:b-qv0t4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L128)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

OK, I added a check on the type of the errors - it is a little hacky though, so if you see a better way to do this I can change it.
</blockquote></details>

I think you can clean this up with either `require.IsType` or `require.ErrorIs`. That should allow you to remove all use f reflection.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 72 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3QBUCb3uBmlXlw6y7S:-N3QBUCb3uBmlXlw6y7T:b-n8qpsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa57796503c630b0ea9c4802d47915c64fc978ea/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L72)):*
> ```Go
> 	// the way that type information is lost, this seemed like the only way to do
> 	// this.
> 	Success := reflect.TypeOf(1)
> ```

nit: local variables are rarely capitalized in Go. Capitalization is used to control package visibility.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 128 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N35MH7z5OwwVoiAmwVt:-N3QNN938SyzU0ZFdHJj:bbv0nue) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c895dfe96a01e7491b148c55007f2a06463a32e/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L128)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think you can clean this up with either `require.IsType` or `require.ErrorIs`. That should allow you to remove all use f reflection.
</blockquote></details>

OK, I was able to get this working with IsType - It was a little different than I thought at first but glad you pushed me down this path. 
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent_test.go` line 72 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-N3QBUCb3uBmlXlw6y7S:-N3QNJkXBNerkIO5-aNc:bm4ypj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa57796503c630b0ea9c4802d47915c64fc978ea/pkg/kv/kvserver/batcheval/cmd_query_intent_test.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: local variables are rarely capitalized in Go. Capitalization is used to control package visibility.
</blockquote></details>

OK, fixed


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81830#-:-N3Q_GLvDRcYArKEABVx:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)



<!-- Sent from Reviewable.io -->

andrewbaptist commented:
bors r=nvanbenschoten

andrewbaptist commented:
bors r=nvanbenschoten
